lib/srv/server/watcher.go

```diff
@@ -51,43 +45,123 @@ type Fetcher interface {
 	IntegrationName() string
 }
 
+// Option is a functional option for the Watcher.
+type Option[Instances any] func(*Watcher[Instances])
+
+// WithPollInterval sets the interval at which the watcher will fetch
+// instances from AWS.
+func WithPollInterval[Instances any](interval time.Duration) Option[Instances] {
+	return func(w *Watcher[Instances]) {
+		w.pollInterval = interval
+	}
+}
+
 // WithTriggerFetchC sets a poll trigger to manual start a resource polling.
-func WithTriggerFetchC(triggerFetchC <-chan struct{}) Option {
-	return func(w *Watcher) {
+func WithTriggerFetchC[Instances any](triggerFetchC <-chan struct{}) Option[Instances] {
+	return func(w *Watcher[Instances]) {
 		w.triggerFetchC = triggerFetchC
 	}
 }
 
 // WithPreFetchHookFn sets a function that gets called before each new iteration.
-func WithPreFetchHookFn(f func()) Option {
-	return func(w *Watcher) {
+func WithPreFetchHookFn[Instances any](f func(fetchers []Fetcher[Instances])) Option[Instances] {
+	return func(w *Watcher[Instances]) {
 		w.preFetchHookFn = f
 	}
 }
 
 // WithClock sets a clock that is used to periodically fetch new resources.
-func WithClock(clock clockwork.Clock) Option {
-	return func(w *Watcher) {
+func WithClock[Instances any](clock clockwork.Clock) Option[Instances] {
+	return func(w *Watcher[Instances]) {
 		w.clock = clock
 	}
 }
 
+// WithMissedRotation sets the missed rotation channel.
+// Specialized for EC2Instances since this functionality is specific to EC2 servers.
+func WithMissedRotation(missedRotation <-chan []types.Server) Option[EC2Instances] {
+	return func(w *Watcher[EC2Instances]) {
+		w.missedRotation = missedRotation
+	}
+}
+
+type syncMap[T any] struct {
```

Comment:
Could we leverage [utils.SyncMap](https://github.com/gravitational/teleport/blob/master/lib/utils/sync_map.go#L27) instead? 

The largest difference in the API seems to be getAll vs Range, though it seems like most callers of getAll immediately iterate over the items to perform some set of work.